src/vma/dev/ib_ctx_handler.cpp

```diff
@@ -53,29 +55,77 @@
 #define ibch_logfuncall         __log_info_funcall
 
 
-
-ib_ctx_handler::ib_ctx_handler(struct ibv_context* ctx, ts_conversion_mode_t ctx_time_converter_mode) :
+ib_ctx_handler::ib_ctx_handler(struct ibv_context* ctx, ts_conversion_mode_t m_p_ctx_time_converter_mode) :
 	m_flow_tag_enabled(false)
 	, m_removed(false)
 	, m_conf_attr_rx_num_wre(0)
 	, m_conf_attr_tx_num_to_signal(0)
 	, m_conf_attr_tx_max_inline(0)
 	, m_conf_attr_tx_num_wre(0)
-	, ctx_time_converter(ctx, ctx_time_converter_mode)
+	, m_p_ctx_time_converter(NULL)
 {
 	memset(&m_ibv_port_attr, 0, sizeof(m_ibv_port_attr));
 	m_p_ibv_context = ctx;
         m_p_ibv_device = ctx->device;
 
         BULLSEYE_EXCLUDE_BLOCK_START
-	if (m_p_ibv_device == NULL)
+	if (m_p_ibv_device == NULL) {
 		ibch_logpanic("ibv_device is NULL! (ibv context %p)", m_p_ibv_context);
+	}
+
+#ifdef DEFINED_IBV_EXP_CQ_TIMESTAMP
+	if (m_p_ctx_time_converter_mode != TS_CONVERSION_MODE_DISABLE) {
+		struct ibv_exp_device_attr device_attr;
+		memset(&device_attr, 0, sizeof(device_attr));
+		device_attr.comp_mask = IBV_EXP_DEVICE_ATTR_WITH_HCA_CORE_CLOCK;
+		if (!ibv_exp_query_device(m_p_ibv_context ,&device_attr)) {
+			if (m_p_ctx_time_converter_mode == TS_CONVERSION_MODE_PTP) {
+#ifdef DEFINED_IBV_EXP_VALUES_CLOCK_INFO
+				struct ibv_exp_values ibv_exp_values_tmp;
+				memset(&ibv_exp_values_tmp, 0, sizeof(ibv_exp_values_tmp));
+				ibv_exp_values_tmp.comp_mask = IBV_EXP_VALUES_CLOCK_INFO;
+				int ret = ibv_exp_query_values(m_p_ibv_context, IBV_EXP_VALUES_CLOCK_INFO, &ibv_exp_values_tmp);
+				if (!ret) {
+					m_p_ctx_time_converter = new time_converter_ptp(ctx);
+				} else { // revert to mode TS_CONVERSION_MODE_SYNC
+					m_p_ctx_time_converter = new time_converter_ib_ctx(ctx, TS_CONVERSION_MODE_SYNC, device_attr.hca_core_clock);
+					ibch_logwarn("ibv_exp_query_values failure for clock_info, reverting to mode TS_CONVERSION_MODE_SYNC (ibv context %p) (return value=%d)",
+								m_p_ibv_context, ret);
+				}
+#else
+				m_p_ctx_time_converter = new ib_ctx_time_converter(ctx, TS_CONVERSION_MODE_SYNC, device_attr.hca_clock);
```

Comment:
You're using the old `ib_ctx_time_converter` again instead of the new `time_converter_ib_ctx`.